timeutil/schedule.go

```diff
@@ -233,3 +514,302 @@ func ParseSchedule(scheduleSpec string) ([]*Schedule, error) {
+func parseSpan(spec string) (string, string) {
+	var start, end string
+
+	specs := strings.Split(spec, spanToken)
+
+	if len(specs) == 1 {
+		start = specs[0]
+		end = start
+	} else {
+		start = specs[0]
+		end = specs[1]
+	}
+	return start, end
+}
+
+// parseTime parses a time specification which can either be `<hh>:<mm>` or
+// `<hh>:<mm>-<hh>:<mm>`. Returns corresponding TimeOfDay structs.
+func parseTime(start, end string) (TimeOfDay, TimeOfDay, error) {
+	// is it a time?
+	var err error
+	var tstart, tend TimeOfDay
+
+	if start == end {
+		// single time, eg. 10:00
+		tstart, err = ParseTime(start)
+		if err == nil {
+			tend = tstart
+		}
+	} else {
+		// time span, eg. 10:00-12:00
+		tstart, tend, err = parseTimeInterval(fmt.Sprintf("%s-%s", start, end))
+	}
+
+	return tstart, tend, err
+}
+
+func parseTimeSpan(start, end string) (*TimeSpan, error) {
+
+	startTime, endTime, err := parseTime(start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	span := &TimeSpan{
+		Start: startTime,
+	}
+	if endTime != startTime {
+		span.End = endTime
+	}
+	return span, nil
+}
+
+// parseWeekday will parse a string and extract weekday (eg. wed),
+// MonthWeekday (eg. 1st, 5th in the month).
+func parseWeekday(s string) (*Weekday, error) {
+	l := len(s)
+	if l != 3 && l != 4 {
+		return nil, fmt.Errorf("cannot parse %q: invalid format", s)
+	}
+
+	day := s
+	var pos uint
+	if l == 4 {
+		day = s[0:3]
+		if v, err := strconv.ParseUint(s[3:], 10, 32); err != nil {
+			return nil, fmt.Errorf("cannot parse %q: invalid week", s)
+		} else if v < 1 || v > 5 {
+			return nil, fmt.Errorf("cannot parse %q: incorrect week number", s)
+		} else {
+			pos = uint(v)
+		}
+	}
+
+	if !IsValidWeekday(day) {
+		return nil, fmt.Errorf("cannot parse %q: invalid weekday", s)
+	}
+
+	week := &Weekday{
+		Day: day,
+		Pos: pos,
+	}
+	return week, nil
+}
+
+// parseCount will parse the string containing a count token and return the
+// count, the rest of the string with count information removed or an error
+func parseCount(s string) (count uint, rest string, err error) {
+	if strings.Contains(s, countToken) {
+		//     timespan = time ( "-" / "~" ) time [ "/" ( time / count ) ]
+		ws := strings.Split(s, countToken)
+		rest = ws[0]
+		countStr := ws[1]
+		c, err := strconv.ParseUint(countStr, 10, 32)
+		if err != nil || c == 0 {
+			return 0, "", fmt.Errorf("cannot parse %q: not a valid event interval",
+				s)
+		}
+		return uint(c), rest, nil
+	}
+	return 0, s, nil
+}
+
+// isValidWeekdaySpec returns true if string is of the following format:
+//     wday =  ( "sun" / "mon" / "tue" / "wed" / "thu" / "fri" / "sat" ) [ DIGIT ]
+func isValidWeekdaySpec(s string) bool {
+	_, err := parseWeekday(s)
+	return err == nil
+}
+
+const (
+	spanToken           = "-"
+	randomizedSpanToken = "~"
+	countToken          = "/"
+)
+
+var specialTokens = map[string]string{
+	// shorthand variant of whole day
+	"-": "0:00-24:00",
```

Comment:
Are those special tokens tested?

Reply:
There are tests for them right here: https://github.com/snapcore/snapd/pull/4313/commits/92325cfada9f6c53bd6eae917eb8f4e6f56a670f#diff-9a58566663280b9c066e815135f62b64R476